lib/perl/Genome/Model/Tools/Speedseq/Base.pm

```diff
@@ -0,0 +1,290 @@
+            is_many => 1,
+        },
+        input_files => {
+            is_many => 1,
+        },
+    ],
+};
+
+sub help_detail {
+    "A flexible framework for rapid genome analysis and interpretation"
+}
+
+my %TOOL_VERSIONS = (
+    'test'     => '/gscmnt/gc2719/halllab/users/cchiang/bin/speedseq',
+);
+
+sub available_versions {
+    my $self = shift;
+    return keys(%TOOL_VERSIONS);
+}
+
+sub path_for_version {
+    my ($class, $version) = @_;
+    unless ($version) {
+        die('No version defined!');
+    }
+    my $path = $TOOL_VERSIONS{$version};
+    return $path if defined $path;
+    die 'No path found for version: '. $version;
+}
+
+sub tool_path {
+    my $self = shift;
+    return $self->path_for_version($self->version);
+}
+
+# tool subcommand name as a string (e.g., 'realign')
+sub _tool_subcommand_name {
+    confess "sub _tool_subcommand_name must be overridden in child class";
+}
+
+sub _resolve_output_files {
+    # example:
+}
+
+sub _resolve_input_files {
+    my $self = shift;
+
+    my @metas = $self->_tool_input_file_metas;
+
+    my @input_files = map { $self->_tool_input_files_from_meta($_) } @metas;
+    $self->input_files(\@input_files);
+    return @input_files;
+}
+
+# tool params should be 'params' and also have tool_param_name defined
+sub _tool_param_metas {
+    my $class = shift;
+    my @param_property_metas = $class->__meta__->properties(is_param => 1);
+    return grep {$_->can("tool_param_name") && $_->tool_param_name} @param_property_metas;
+}
+
+# tool inputs should be 'inputs' and also have tool_input_bare_arg_position defined
+sub _tool_input_metas {
+    my $class = shift;
+    my @input_property_metas = $class->__meta__->properties(is_input => 1);
+    my @tool_input_metas = grep {$_->can("tool_bare_arg_position") && $_->tool_bare_arg_position} @input_property_metas;
+    return sort { $a->tool_bare_arg_position <=> $b->tool_bare_arg_position } @tool_input_metas;
+}
+
+sub _tool_input_file_metas {
+    my $class = shift;
+    my @property_metas = $class->__meta__->properties();
+    return grep {$_->can("tool_input_file") && $_->tool_input_file} @property_metas;
+}
+
+# Handle conversion from boolean => 'true'/'false' and general tool arg
+# formatting (argname=argvalue).
+sub _format_tool_arg {
+    my ($type, $name, $value) = @_;
+    return '-'. $name if ($type eq 'Boolean');
+    $value = qq{"$value"} if $type eq 'Text';
+    return sprintf '-%s %s', $name, $value;
+}
+
+# given a property meta object (defining property name, and ostensibly
+# tool_param_name), return the list of cmdline args that should be passed
+# to the tool based on the values set on the current object.
+#
+# this is where is_many properties get handled.
+sub _tool_param_args_from_meta {
+    my ($self, $meta) = @_;
+
+    my $ur_name = $meta->property_name;
+    # using // instead of ? : below made a lims perl test fail
+    my $type = defined $meta->data_type ? $meta->data_type : 'String';
```

Comment:
Again, I elected to leave this as similar to the Picard code so it could be merged at some point.  I think the intent of 'Text' type was to quote it since it could be a filesystem path.  'String' would not be quoted.